R/cache.R

```diff
@@ -0,0 +1,213 @@
+#'   logfile = "some/subdir/logs.txt",
+#'   prune_rate = 20L
+#' )
+#' }
+#'
+#' @inheritParams set_cache
+#' @export
+clear_cache <- function(...) {
+  cache_environ$epidatr_cache$destroy()
+  set_cache(...)
+}
+
+#' turn off the caching for this session
+#' @description
+#' Disable caching until you call `set_cache` or restart R. The files defining the cache are untouched. If you are looking to disable the caching more permanently, set `EPIDATR_USE_CACHE=FALSE` as environmental variable in your `.Renviron`.
+#' @export
+disable_cache <- function() {
+  cache_environ$epidatr_cache <- NULL
+}
+
+#' turn off the caching for this session
+#' @description
+#' Print out the information about the cache (as would be returned by cachem's `info()` method)
+#' @export
+cache_info <- function() {
+  cache_environ$epidatr_cache$info()
+}
+
+#' create a new cache for this session
+#'
+#' @description
+#' the guts of caching, its interposed between fetch and the specific fetch methods. Internal method only.
+#'
+#' @param call the `epidata_call` object
+#' @inheritParams fetch
+#' @import cachem openssl
+cache_epidata_call <- function(epidata_call, fetch_args = fetch_args_list()) {
+  as_of_cachable <- (!is.null(epidata_call$params$as_of) && epidata_call$params$as_of != "*")
+  issues_cachable <- (!is.null(epidata_call$params$issues) && epidata_call$params$issues != "*")
+  is_cachable <- (
+    cache_environ$use_cache &&
+      !is.null(cache_environ$epidatr_cache) &&
+      (as_of_cachable || issues_cachable) &&
+      !(fetch_args$dry_run) &&
+      is.null(fetch_args$base_url) &&
+      !fetch_args$debug &&
+      fetch_args$format_type != "json"
+  )
+  if (is_cachable) {
+    target <- request_url(epidata_call)
+    hashed <- md5(target)
+    cached <- cache_environ$epidatr_cache$get(hashed)
+    if (!is.key_missing(cached)) {
+      cli::cli_warn("loading from the cache at {cache_environ$epidatr_cache$info()$dir}; see {cache_environ$epidatr_cache$info()$logfile} for more details.",
+        .frequency = "regularly",
+        .frequency_id = "using the cache"
+      )
+      return(cached[[1]])
+    }
+  }
+  'which was saved on {format(cached[[2]],"%A %B %d, %Y")}, which took {round(cached[[3]][[3]], digits=5)} seconds.'
+  # need to actually get the data, since its either not in the cache or we're not caching
+  runtime <- system.time(if (epidata_call$only_supports_classic) {
+    fetched <- fetch_classic(epidata_call, fetch_args)
+  } else {
+    fetched <- fetch_tbl(epidata_call, fetch_args)
+  })
+  # add it to the cache if appropriate
+  if (is_cachable) {
+    cache_environ$epidatr_cache$set(hashed, list(fetched, Sys.time(), runtime))
+  }
+  return(fetched)
+}
+
+# debugonce(cache_epidata_call)
+# thing <- covidcast(
+#   source = "jhu-csse",
+#   signals = "confirmed_7dav_incidence_prop",
+#   geo_type = "state",
+#   time_type = "day",
+#   geo_values = "ca,fl",
+#   time_values = epirange(20200601, 20200801),
+#   as_of = "20230101",
+# )
```

Comment:
```suggestion
```

Reply:
surprised that made it into a commit... oops. Guess I need to double check I've run the styler before all this is through.